From patchwork Mon Jan 28 18:12:51 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10784295
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CBA961390
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 18:13:04 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id B9BEA2BF2F
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 18:13:04 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id ADFBB2BFC6; Mon, 28 Jan 2019 18:13:04 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 431F72BF2F
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 18:13:04 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727163AbfA1SND (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 28 Jan 2019 13:13:03 -0500
Received: from mail-pg1-f202.google.com ([209.85.215.202]:56246 "EHLO
        mail-pg1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727143AbfA1SND (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 28 Jan 2019 13:13:03 -0500
Received: by mail-pg1-f202.google.com with SMTP id g188so11982413pgc.22
        for <kvm@vger.kernel.org>; Mon, 28 Jan 2019 10:13:02 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=dAs38xTzfHfMScBcpOObXDFHAmFen4dZvsnl7CemAD4=;
        b=S79Woq86jeeOD5jwTubdOGp/tn/5cFGUNp3LawcFDB+516u8gWyolryIOYOMg0128g
         w+wqnEBd6h/DPVAEAz/GIbtMMv4w+DyJB51B3s6p/Wa3GnfsWHnZh9LGGwmwXpcO8hxi
         enBj/09hbnrgc8eCxrzFXPJp9qHWfVjzfuTmhcquhPegOFqEPgZ6M7KKF4VovD+rBMgx
         +dqqKkuPXeF2HIsyebqPIEcD+Oz7nBlw5rIpicOhIREaKr5/l+QIELJ8zO0DA2GR87Pn
         2jRDz8KYlzKg/bdlLUctEwJ5Gsb0zVZr0EX7nl615Hm/6X467K9CgChIbblz+r1f4/7I
         uKiw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=dAs38xTzfHfMScBcpOObXDFHAmFen4dZvsnl7CemAD4=;
        b=L+KUgCzPJBenbhTJprOBQTS/WoOPgfQtKCR1//Uk8Ey8xYipjXdm/QNvfhFhqsEPBm
         GbsoURouZ+qH96wUFQLBHpPjoIdVTAIVWKnugWKiMm6S4SBfe7w9UxykF40744HTXTwd
         fw62M/DQa4klvnr3NPkXcFNKSyezRf6OgIetgLVQTmMgs+SbSDhT+6eME0vDxiswf7iW
         1YF1as4Gz4ArHhVm77hbawCdZoD+Ohy9Z52H1C0DeG0AAC15yAvTMiuXkuTB9OdfKM7v
         cY/LMs42235VVO9pZLFYp95vZRwjmZHgbs6r1qqmxdrNIO1W3zorUvd4fIPpMA92pBVi
         8baw==
X-Gm-Message-State: AHQUAuZ8NX6gnRVEOw4Djg+tlkS0n3avvNP/H+SnUGvMzrwpGvC7BsCO
        rDpCCqM7NQtzW2KdPy0NxjuF0l3mH2Fg5W9sdzoLgkOTt41JQI+zs/Z+egFHHWbOW7/f+0Gghtw
        veOGXgDnS9vvWC6sceDQTYcqQ3UUPxm9lJ0UJ1IOFWQqGq8w0JVbGEAYfUqktqyA=
X-Google-Smtp-Source: 
 AHgI3IYBqoDlZdAw1ijBsq8MHCIFSEhvy0RmBJW+eRb7p7qfsD6DCCrH2bYBRxhzwzEjmEWR+lGz/ohAVxXReQ==
X-Received: by 2002:a17:902:8f95:: with SMTP id
 z21mr321822plo.99.1548699181921;
 Mon, 28 Jan 2019 10:13:01 -0800 (PST)
Date: Mon, 28 Jan 2019 10:12:51 -0800
Message-Id: <20190128181252.245510-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.20.1.495.gaa96b0ce6b-goog
Subject: [PATCH 1/2] kvm: x86: IA32_ARCH_CAPABILITIES is only emulated for
 Intel
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Marc Orr <marcorr@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

This MSR is not emulated on AMD hosts.

Fixes: 1eaafe91a0df ("kvm: x86: IA32_ARCH_CAPABILITIES is always supported")
Reported-by: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Eduardo Habkost <ehabkost@redhat.com>
Tested-by: Eduardo Habkost <ehabkost@redhat.com>
---
 arch/x86/include/asm/kvm_host.h | 3 ++-
 arch/x86/kvm/cpuid.c            | 7 +------
 arch/x86/kvm/svm.c              | 3 ++-
 arch/x86/kvm/vmx/vmx.c          | 9 ++++++++-
 4 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index 4660ce90de7f..cd84d1bf33d0 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1095,7 +1095,8 @@ struct kvm_x86_ops {
 
 	void (*set_tdp_cr3)(struct kvm_vcpu *vcpu, unsigned long cr3);
 
-	void (*set_supported_cpuid)(u32 func, struct kvm_cpuid_entry2 *entry);
+	void (*set_supported_cpuid)(u32 func, u32 index,
+				    struct kvm_cpuid_entry2 *entry);
 
 	bool (*has_wbinvd_exit)(void);
 
diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index bbffa6c54697..39b92f3409b6 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -496,11 +496,6 @@ static inline int __do_cpuid_ent(struct kvm_cpuid_entry2 *entry, u32 function,
 				entry->ecx &= ~F(PKU);
 			entry->edx &= kvm_cpuid_7_0_edx_x86_features;
 			cpuid_mask(&entry->edx, CPUID_7_EDX);
-			/*
-			 * We emulate ARCH_CAPABILITIES in software even
-			 * if the host doesn't support it.
-			 */
-			entry->edx |= F(ARCH_CAPABILITIES);
 		} else {
 			entry->ebx = 0;
 			entry->ecx = 0;
@@ -718,7 +713,7 @@ static inline int __do_cpuid_ent(struct kvm_cpuid_entry2 *entry, u32 function,
 		break;
 	}
 
-	kvm_x86_ops->set_supported_cpuid(function, entry);
+	kvm_x86_ops->set_supported_cpuid(function, index, entry);
 
 	r = 0;
 
diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index a157ca5b6869..c34a6d77d8e4 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -5882,7 +5882,8 @@ static void svm_cpuid_update(struct kvm_vcpu *vcpu)
 	guest_cpuid_clear(vcpu, X86_FEATURE_X2APIC);
 }
 
-static void svm_set_supported_cpuid(u32 func, struct kvm_cpuid_entry2 *entry)
+static void svm_set_supported_cpuid(u32 func, u32 index,
+				    struct kvm_cpuid_entry2 *entry)
 {
 	switch (func) {
 	case 0x1:
diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index f6915f10e584..18e0975486ee 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -7073,10 +7073,17 @@ static void vmx_cpuid_update(struct kvm_vcpu *vcpu)
 		update_intel_pt_cfg(vcpu);
 }
 
-static void vmx_set_supported_cpuid(u32 func, struct kvm_cpuid_entry2 *entry)
+static void vmx_set_supported_cpuid(u32 func, u32 index,
+				    struct kvm_cpuid_entry2 *entry)
 {
 	if (func == 1 && nested)
 		entry->ecx |= bit(X86_FEATURE_VMX);
+	else if (func == 7 && index == 0)
+		/*
+		 * We emulate ARCH_CAPABILITIES in software even
+		 * if the host doesn't support it.
+		 */
+		entry->edx |= bit(X86_FEATURE_ARCH_CAPABILITIES);
 }
 
 static void vmx_request_immediate_exit(struct kvm_vcpu *vcpu)

From patchwork Mon Jan 28 18:12:52 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10784297
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id E86276C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 18:13:07 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D89372BF07
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 18:13:07 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id CCE792BF41; Mon, 28 Jan 2019 18:13:07 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 79C6E2BF07
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 28 Jan 2019 18:13:07 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727180AbfA1SNG (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 28 Jan 2019 13:13:06 -0500
Received: from mail-pl1-f202.google.com ([209.85.214.202]:43078 "EHLO
        mail-pl1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727172AbfA1SNG (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 28 Jan 2019 13:13:06 -0500
Received: by mail-pl1-f202.google.com with SMTP id g13so12327656plo.10
        for <kvm@vger.kernel.org>; Mon, 28 Jan 2019 10:13:05 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=bLw4lxQ8qY1jWFyj72c1RwdOFvqvBvBMQRS4C6dbpqw=;
        b=gwliC1vJY8g6+TylCHuJ3qWFwd1Z/518PN6ETYXf4gposZ+soEAbhVWsTUPJD5eHOd
         pmevI3UII3qi1Fv4pRmVluSXn9dkOgXHWQ1hE0OAyBDorZEo06H6oVlcGgVCjtS9U0m4
         Br+oM5U4NpYXgKWQNenzzSlLm4ID4BiBq3AhfeeNM16RwWcAksg2uAR3We8knYhe22L3
         n4ZYjkai3YtGU0VzXE7DPjo+9DIh7SELTFzC0fDAWOn1l2BOkrl8KDEDiVrOJhfa3Lvp
         6bXwYIDn47uE0Y/DtDY0lT4JXY8ATPRQVPNVWteIhvoQ3xVzqxbrnOe+JrNGQ/Zyb7Yo
         naAA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=bLw4lxQ8qY1jWFyj72c1RwdOFvqvBvBMQRS4C6dbpqw=;
        b=A0ElR5pKoBLNrJRmtzKP8/svc+FUFMhQu6e7ZpgL+Xnz101yMiB/N86GXgCzTiNsND
         Qyt1Wx6eON7zaEZxWSRj9+SLPdqCgbSvevxLrmIsGWJlJZNlwjzsn+N7Vy32ejvj9+hN
         r44k13g2i3MvfduDMKUqiKKc+WCe4KfNTcp/Up0dubFt2ZWxbUhYyG8EmxrO/cJK43uL
         guG5N3vjc8NkmYbFlWQTjZfiXhJExFGxbn2RSXMGI8s3p/MVMbn0JZw/7nw8SnOBrEQj
         9Ato2lwZ4XHspgZMEO4h7hADbSg6aYafC9CjwtzlbVzpKPzHCUGDR0HATelOaoGIl1oD
         Wk1w==
X-Gm-Message-State: AHQUAub0IxH23L66lsQy/TtMxBS3yY0P3FVfGks4MDoV0Gy+QiIfI/TB
        21letbzvqT+OZRR/wKYpF8R4t8tYNk4l6GvJJLlwHTdJdCOeZa5ClnnTrC/zYN0rsbsrErtA5DS
        wVnSgfhZBI3bftG7TDj0co+uLz2weWGBI/zsvuftvekxXlJ06aLTqPsSDai8NRLA=
X-Google-Smtp-Source: 
 AHgI3IbRONstlOQu1aVvzcfHtww0vWqVaT1jK2p0J6QgvjexeV+1UXosTHJL6hMTCe/T73Fz33NbpFzeMSv7zw==
X-Received: by 2002:a17:902:2a6b:: with SMTP id
 i98mr320986plb.67.1548699185535;
 Mon, 28 Jan 2019 10:13:05 -0800 (PST)
Date: Mon, 28 Jan 2019 10:12:52 -0800
In-Reply-To: <20190128181252.245510-1-jmattson@google.com>
Message-Id: <20190128181252.245510-2-jmattson@google.com>
Mime-Version: 1.0
References: <20190128181252.245510-1-jmattson@google.com>
X-Mailer: git-send-email 2.20.1.495.gaa96b0ce6b-goog
Subject: [PATCH 2/2] kvm: x86: IA32_ARCH_CAPABILITIES may be emulated on Intel
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Marc Orr <marcorr@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

This MSR should be listed in emulated_msrs[] rather than
msrs_to_save[], so that it will always be reported by
KVM_GET_MSR_INDEX_LIST. However, it is not emulated on AMD hosts.

Fixes: 1eaafe91a0df ("kvm: x86: IA32_ARCH_CAPABILITIES is always supported")
Reported-by: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Eduardo Habkost <ehabkost@redhat.com>
Tested-by: Eduardo Habkost <ehabkost@redhat.com>
---
 arch/x86/kvm/svm.c | 1 +
 arch/x86/kvm/x86.c | 3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index c34a6d77d8e4..96b680713301 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -5856,6 +5856,7 @@ static bool svm_has_emulated_msr(int index)
 {
 	switch (index) {
 	case MSR_IA32_MCG_EXT_CTL:
+	case MSR_IA32_ARCH_CAPABILITIES:
 		return false;
 	default:
 		break;
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 02c8e095a239..e47e24bc9500 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -1125,7 +1125,7 @@ static u32 msrs_to_save[] = {
 #endif
 	MSR_IA32_TSC, MSR_IA32_CR_PAT, MSR_VM_HSAVE_PA,
 	MSR_IA32_FEATURE_CONTROL, MSR_IA32_BNDCFGS, MSR_TSC_AUX,
-	MSR_IA32_SPEC_CTRL, MSR_IA32_ARCH_CAPABILITIES,
+	MSR_IA32_SPEC_CTRL,
 	MSR_IA32_RTIT_CTL, MSR_IA32_RTIT_STATUS, MSR_IA32_RTIT_CR3_MATCH,
 	MSR_IA32_RTIT_OUTPUT_BASE, MSR_IA32_RTIT_OUTPUT_MASK,
 	MSR_IA32_RTIT_ADDR0_A, MSR_IA32_RTIT_ADDR0_B,
@@ -1167,6 +1167,7 @@ static u32 emulated_msrs[] = {
 	MSR_PLATFORM_INFO,
 	MSR_MISC_FEATURES_ENABLES,
 	MSR_AMD64_VIRT_SPEC_CTRL,
+	MSR_IA32_ARCH_CAPABILITIES,
 };
 
 static unsigned num_emulated_msrs;
